matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/membership/RoomMemberEventHandler.kt

```diff
@@ -41,10 +60,56 @@ internal class RoomMemberEventHandler @Inject constructor() {
         }
         val roomMemberEntity = RoomMemberEntityFactory.create(roomId, userId, roomMember)
         realm.insertOrUpdate(roomMemberEntity)
-        if (roomMember.membership.isActive()) {
-            val userEntity = UserEntityFactory.create(userId, roomMember)
-            realm.insertOrUpdate(userEntity)
+
+        if (roomMember.membership.isActive() && updateUserEntity(realm, userId, roomMember)) {
+            if (userId == this.userId) {
+                fetchUserProfile(userId) // To fix #1715 (myroomnick changes the global name)
+            }
         }
         return true
     }
+
+    private fun updateUserEntity(realm: Realm, userId: String, roomMember: RoomMemberContent): Boolean {
+        return realm
+                .where<UserEntity>()
+                .equalTo(UserEntityFields.USER_ID, userId)
+                .findFirst()
+                ?.let { userEntity ->
+                    if (userEntity.displayName != roomMember.displayName || userEntity.avatarUrl != roomMember.avatarUrl) {
+                        userEntity.displayName = roomMember.displayName ?: ""
+                        userEntity.avatarUrl = roomMember.avatarUrl ?: ""
+                        return true
+                    }
+                    return false
+                }
+                ?: run {
+                    val userEntity = UserEntityFactory.create(userId, roomMember)
+                    realm.insertOrUpdate(userEntity)
+                    return true
+        }
+    }
+
+    private fun fetchUserProfile(userId: String) {
+        val params = GetProfileInfoTask.Params(userId)
```

Comment:
If you let that here, could you create a task or use awaitCallback instead?

Reply:
Sorry, I didn't understand your suggestion. Could you describe it a bit?